### ISSUE TITLE
Midi Import crash in master branch

### ISSUE THREAD

**Issue by midi-pascal** [LMMS/lmms#1981]
Midi Import crash when importing some files in master branch.
Version 1.0 does not crash when importing the same files.

The bug lies within the MidiImport::readSMF method when tracing unhandled Midi events.
The crash occurs there:
`				if(!handled) {`
`					printf("MISSING GLOBAL THINGY\n");`
`					printf("     %d %d %f %s %s\n", (int) evt->chan,`
`					       evt->get_type_code(), evt->time,`
`					       evt->get_attribute(), evt->get_atom_value() );`
`					// Global stuff`
`				}`

It looks like at least one Midi event type is not decoded properly.
Also - if I remember well - we are not supposed to use printf() to make a trace, uh?
Actually I try to solve this issue.
More to come :-)

**Comment by softrabbit** [comment]
I filed a pull request for this bug yesterday, so it should be fixed soon-ish in master.

**Comment by midi-pascal** [comment]
Ha!
Thanks a lot.
I can stop investigating this issue...

**Comment by michaelgregorius** [comment]
Hi @softrabbit,

you beat me by about half an hour. :smile: I have a fix this bug as well (my first pull request):
https://github.com/LMMS/lmms/pull/1984

I fixed it a bit differently by adding checks for the conditions which are asserted during get_atom_value. How can I attach my example file that I have used to reproduce the crash to this issue?

**Comment by softrabbit** [comment]
@michaelgregorius, I don't think you can attach anything that's not an image, it has to be made available somewhere else. I wonder if one could create a (self-extracting) archive of some kind in a valid PNG file?

**Comment by tresf** [comment]
> I wonder if one could create a (self-extracting) archive of some kind in a valid PNG file?

:laughing: ^---- Voodoo magic... :)

There's actually quite a few people doing this, but most of them involve text from what I'm reading.... :)

**Comment by Spekular** [comment]
Encode the png in an audio file, then place the audio file in a game and
require players to bring radios to certain locations to hear it. :+1:

**Comment by mikobuntu** [comment]
You can easily embed an audio file into a .gif by using " cat some.gif some.mp3 > another.gif " but for easy extraction I'm unsure.

**Comment by Umcaruje** [comment]
@michaelgregorius can you confirm that this was fixed with #1984?

**Comment by michaelgregorius** [comment]
@Umcaruje No, unfortunately I cannot confirm with 100% certainty that #1984 fixes the issue described in this ticket. I was only able to reproduce a crash using one of my own MIDI files. To ensure that this issue is fixed by #1984 we'd need one of the files that @midi-pascal used.

**Comment by midi-pascal** [comment]
I will build and test rigth now and let you know if the problem is fixed 
in a few minutes.

**Comment by midi-pascal** [comment]
Hi,

I updated from Git and rebuild lmms from scratch.
As far as can see, the issue is fixed.
I imported my two killer Midi files with no crash.

FYI, the crash was occuring with Midi files that contain lyrics - kind 
of karaoke Midi files - when decoding the lyrics track.
In MidiImport::readSMF(), evt->get_attribute() returns "tracknames" but 
evt->get_update_type() returns "u" instead of "a".
Not sure why however.

Thanks a lot for the fix anyways, guys :-)
